### ISSUE TITLE
Anonymous controller can't access application routes

### ISSUE THREAD

**Issue by n0nick** [rspec/rspec-rails#1321]
This is in fact the exact behavior described in issue #573.
I was able to reproduce it in version 3.2.0 of the `rspec` and `rspec-rails` gems.

Seems like the application routes are not accessible to controller code when it runs through RSpec.

See this gist for my reproduction code and results:
https://gist.github.com/n0nick/594432eee7ccf7ee2a2c

**Comment by cupakromer** [comment]
This doesn't appear to be something that is simple to implement. To my knowledge, controllers can only have a single route set.

The current route set used is the application routes, which makes all of the app routes and helpers available. It also makes the common `redirect_to model` or `redirect_to sign_in_url` available. Which tend to appear regularly in `ApplicationController` classes.

**Comment by cupakromer** [comment]
To clarify the application routes are available. It is the custom anonymous controller routes which are not.

**Comment by n0nick** [comment]
@cupakromer that is correct - what I'm missing is access to the spec's anonymous controller routes from within the actual controller's code.

What's strange to me is that this seems to have been fixed when #573 was closed, i.e. in commit 648ba097. From what I'm seeing now, the code introduced in that commit is no longer there, but I didn't yet have time to bisect the history and find out where it was removed.

The workaround I found for this, described in the original #573, was to do
```ruby
  before { controller.stub(:_routes).and_return(@routes) }
```

I'm not entirely sure where `@routes` comes from in controller specs, but it seems to include both the application routes and the custom anonymous controller routes. So this does indicate that the fix introduced in 648ba097 should still be valid.

**Comment by cupakromer** [comment]
> 'm not entirely sure where `@routes` comes from in controller specs, but it seems to include both the application routes and the custom anonymous controller routes.

That instance variable only holds the routes for the anonymous controller. There is a second private instance variable `@orig_routes` which holds the application routes.

The controller spec implements some method missing logic to check both route sets in the spec. This allows you to have access to route helpers like `widgets_path` AND still use `get :index` for the anonymous controller.

**Comment by sshaw** [comment]
Having this problem with 3.3.2. The scary `before` hook fix does not work.

**Comment by jcoyne** [comment]
In rspec rails 3.5.1 I'm able to work around the problem by doing:
```ruby
before { allow(controller).to receive(:_routes).and_return(Rails.application.routes) }
```

**Comment by camertron** [comment]
Any update on this guy?
